widget/embedded/src/store/slices/data.ts

```diff
@@ -127,31 +136,65 @@ export const createDataSlice: StateCreator<
 
         return true;
       })
-      .sort((a, b) => {
-        // Check pinned tokens
-        if (get().isTokenPinned(a, options.type)) {
-          return -1;
+      .sort((tokenA, tokenB) => {
+        //1. sort by pinned tokens
+        const isToken1Pinned = get().isTokenPinned(tokenA, options.type);
+        const isToken2Pinned = get().isTokenPinned(tokenB, options.type);
+        if (isToken1Pinned !== isToken2Pinned) {
+          return isToken1Pinned ? -1 : 1;
+        }
+
+        //2. sort by balance
+        if (options.getBalanceFor) {
+          const token1Balance = options.getBalanceFor(tokenA);
+          const token2Balance = options.getBalanceFor(tokenB);
+          const balanceCompare = compareTokenBalance(
+            token1Balance,
+            token2Balance
+          );
+          if (balanceCompare !== 0) {
+            return balanceCompare;
+          }
         }
-        if (get().isTokenPinned(b, options.type)) {
-          return 1;
+
+        //3. sort by token popularity
+        if (tokenA.isPopular !== tokenB.isPopular) {
+          return tokenA.isPopular ? -1 : 1;
         }
 
-        // Check secondary coins
-        if (a.isSecondaryCoin) {
-          return 1;
+        //4. sort by search phrase
+        if (options.searchFor) {
+          const symbolSearchForCompare = compareWithSearchFor(
+            tokenA,
+            tokenB,
+            options.searchFor
+          );
+          if (symbolSearchForCompare) {
+            return symbolSearchForCompare;
+          }
         }
-        if (b.isSecondaryCoin) {
-          return -1;
+
+        const blockChainA = blockchainNameMap[tokenA.blockchain];
+        const blockChainB = blockchainNameMap[tokenB.blockchain];
+
+        //5. sort by native token
+        const isNativeTokenA = tokenIsNative(tokenA, blockChainA);
+        const isNativeTokenB = tokenIsNative(tokenB, blockChainB);
+        if (isNativeTokenA !== isNativeTokenB) {
+          return isNativeTokenA ? -1 : 1;
         }
 
-        // Check it has an address or not.
-        if (!a.address && b.address) {
-          return -1;
+        //6. sort by token isSecondary
+        if (tokenA.isSecondaryCoin !== tokenB.isSecondaryCoin) {
+          return tokenA.isSecondaryCoin ? 1 : -1;
         }
-        if (a.address && !b.address) {
-          return 1;
+
+        //7. sort by blockchain order
+        if (!!blockChainA && !!blockChainB) {
+          return blockChainA.sort - blockChainB.sort;
         }
 
+        // return for equality
```

Comment:
You can remove this comment. that's how `.sort` works in JS.

Reply:
Fixed